### DISCUSSION TITLE
How to Use Transitions with Windows?

### DISCUSSION THREAD

**Discussion by fitdev** [AvaloniaUI/Avalonia#11817] [Q&A] [answered]
I was trying to add a fade-in effect for a splash window using `Opacity` property and `DoubleTransition` but ti did not work - there was no smooth transition - just an immediate change. Tried it with `Width` too with the same result. Yet it works just fine on for example a `Button` control.

Is that a limitation of Avalonia, or am I doing something wrong? Using latest RC.

```xaml
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
        x:Class="AvaloniaApplication5.MainWindow"
        Title="AvaloniaApplication5"
        Width="500">
  <Window.Transitions>
    <Transitions>
      <DoubleTransition Property="Width" Duration="0:0:1.5"/>
    </Transitions>
  </Window.Transitions>
  <StackPanel Background="Red">
    <TextBlock Background="Blue">Foo</TextBlock>
    <Button x:Name="btn1" Click="btnwidth" Width="400" Background="Green">
      <Button.Transitions>
        <Transitions>
          <DoubleTransition Property="Width" Duration="0:0:1.5"/>
        </Transitions>
      </Button.Transitions>
      <Button.Content>
        Change Width
      </Button.Content>
    </Button>
  </StackPanel>
</Window>
```

```cs
using Avalonia.Controls;
using Avalonia.Interactivity;

namespace AvaloniaApplication5;
public partial class MainWindow : Window {
  public MainWindow() {
    InitializeComponent();
  }
  public void btnwidth(object sender, RoutedEventArgs e) {
    Width = 100; //Does not work
    ;
    //btn1.Width = 100; //Works
  }

}
```

**Comment by timunie** [comment]
I guess this doesn't work well for Windows as these are handled by the OS mainly. I'd use a fully transparent window and fade in the content instead.

**Reply by fitdev** [reply]
Thanks for helping out! But how would I do that, if I set window's `Opacity` to `0`, wouldn't that make content invisible as well?

**Reply by fitdev** [reply]
Also, then I think it should be noted in the docs on Transitions that most of them won't work on windows.

**Reply by timunie** [reply, marked as answer]
```xml
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x='http://schemas.microsoft.com/winfx/2006/xaml'
        TransparencyLevelHint="Transparent"
        Background="Transparent"
        SystemDecorations="None"
        x:Class="Sandbox.MainWindow">
  <StackPanel x:Name="root" Background="Red" Opacity="0" >
    <StackPanel.Transitions>
      <Transitions>
        <DoubleTransition Property="Opacity" Duration="0:0:1.5"/>
      </Transitions>
    </StackPanel.Transitions>
    <TextBlock Background="Blue">Foo</TextBlock>
    <Button Width="400" Background="Green">
      <Button.Content>
        Random Button
      </Button.Content>
    </Button>
  </StackPanel>
</Window>
```

```cs
public partial class MainWindow : Window
{
    public MainWindow()
    {
        InitializeComponent();
    }

    protected override void OnLoaded()
    {
        base.OnLoaded();
        root.Opacity = 1;
    }
}
```

This gives me a nice fade in animation.

**Reply by fitdev** [reply]
Thanks so much for your help!

**Reply by fitdev** [reply]
Just tried your suggestion on Windows 11 and it does not seem to work unfortunately:

```cs
public class SplashScreen : UserControl {

  public SplashScreen() {
    Opacity = 0;
    Transitions = new Transitions {
      new DoubleTransition() { Property = Control.OpacityProperty, Duration = TimeSpan.FromSeconds(1.5) }
    };
  }
}

public class SplashWindow : Window {

  public SplashWindow(SplashScreen screen) {
    WindowStartupLocation = WindowStartupLocation.CenterScreen;
    CanResize = false;
    SystemDecorations = SystemDecorations.None;
    ShowActivated = false;
    ShowInTaskbar = false;
    WindowState = WindowState.Normal;
    Background = Brushes.Transparent;
    TransparencyLevelHint = new[] { WindowTransparencyLevel.Transparent };
    ExtendClientAreaChromeHints = Avalonia.Platform.ExtendClientAreaChromeHints.SystemChrome; //Shadow on Win11 around window
    Content = screen;
  }

  protected override void OnLoaded() {
    base.OnLoaded();
    ((SplashScreen)Content!).Opacity = 1;
  }
}

```

Regardless of what I use for `ExtendClientAreaChromeHints` or `Background` or initial `Opacity`, I still only get fixed opacity.

**Reply by timunie** [reply]
in debug or release mode? if in release mode, make sure to build it with modifier win10-x64 iirc.

**Reply by fitdev** [reply]
I have not tried publishing with RID yet. Though in both Debug and Release modes it does not work, which can perhaps be addressed sometime in the future as it would be nice to see real experience even when debugging without having to publish first.

I also tried modifying app.manifest file, which does get picked up by Windows, but once again seems to have no effect.

Thanks for your suggestion. I will try publishing and let you know.

**Reply by timunie** [reply]
tried to reproduce your code, but tbh I'm so used to XAML that I can't solve your code. Sorry, but good luck anyway.

**Reply by fitdev** [reply]
Thank you for helping out! Really appreciate your effort!

These are just some helper classes which I was hoping to use to create a concrete Splash Screen implementation (app-specific Splash Screen would inherit from `SplashScreen` instead of `UserControl` and then be hosted inside of my `SplashWindow` as its sole content, hence I could use you suggestion of applying transitions to window's content instead of the window itself.

Plus as a bonus that way, for mobile, I could use the same splash screen but hosted in a non-window container (obviously opacity transition would not longer make sense there, but at least I get to reuse splash UI).

**Reply by fitdev** [reply]
Here's a repro of Splash Opacity Transition not working.
[AvaloniaSplashOpacityEffectSample.zip](https://github.com/AvaloniaUI/Avalonia/files/11816232/AvaloniaSplashOpacityEffectSample.zip)
